test/test_sync.cpp

```diff
@@ -4892,6 +4892,185 @@ TEST(Sync_MergeLargeChangesets)
+        table->add_column(type_Int, "integer column");
+        wt.commit();
+    }
+
+    {
+        WriteTransaction wt(db_2);
+        TableRef table = wt.add_table("class_table name");
+        table->add_column(type_Int, "integer column");
+        wt.commit();
+    }
+
+    {
+        for (int i = 0; i < number_of_changesets; ++i) {
+            WriteTransaction wt(db_1);
+            TableRef table = wt.get_table("class_table name");
+            for (int j = 0; j < number_of_instructions; ++j) {
+                auto obj = table->create_object();
+                obj.set("integer column", 2 * j);
+            }
+            wt.commit();
+        }
+    }
+
+    {
+        for (int i = 0; i < number_of_changesets; ++i) {
+            WriteTransaction wt(db_2);
+            TableRef table = wt.get_table("class_table name");
+            for (int j = 0; j < number_of_instructions; ++j) {
+                auto obj = table->create_object();
+                obj.set("integer column", 2 * j + 1);
+            }
+            wt.commit();
+        }
+    }
+
+    {
+        TEST_DIR(dir);
+        MultiClientServerFixture fixture(2, 1, dir, test_context);
+        std::thread th;
+        version_type user_commit_version;
+
+        Session::Config config;
+        config.split_remote_changesets = true;
+        config.on_before_download_integrated = [&](size_t num_changesets) {
+            if (num_changesets == 0) {
+                return;
+            }
+
+            th = std::thread([&] {
+                auto write = db_1->start_write();
+                // Keep the transaction open until the sync client is waiting to acquire the write lock.
+                while (!db_1->waiting_for_write_lock()) {
+                    millisleep(1);
+                }
+                write->close();
+
+                // Sync client holds the write lock now, so commit a local change.
+                WriteTransaction wt(db_1);
+                TableRef table = wt.get_table("class_table name");
+                auto obj = table->create_object();
+                obj.set("integer column", 100000);
+                user_commit_version = wt.commit();
+            });
+        };
+        Session session_1 = fixture.make_session(0, db_1, std::move(config));
+        fixture.bind_session(session_1, 0, "/test");
+        Session session_2 = fixture.make_session(1, db_2);
+        fixture.bind_session(session_2, 0, "/test");
+
+        fixture.start();
+
+        session_1.wait_for_upload_complete_or_client_stopped();
+        session_2.wait_for_upload_complete_or_client_stopped();
+
+        // Wait to download and integrate the remote changes.
+        session_1.wait_for_download_complete_or_client_stopped();
+
+        th.join();
+        // Check that local change was commited before all remote changes were integrated.
+        ReadTransaction rt(db_1);
+        CHECK(user_commit_version < rt.get_version());
```

Comment:
Ah we check for it here. Seems like this could be racy though?

Reply:
Never seen it in practice, but it could lead to a race. I'll think of something better.